server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/GitRepositoryV2.java

```diff
@@ -0,0 +1,874 @@
+        final InternalRepository repo = secondaryRepo != null ? secondaryRepo : primaryRepo;
+        return shouldCreateRollingRepository(repo.headRevision(), minRetentionCommits, minRetentionDays);
+    }
+
+    @Nullable
+    private Revision shouldCreateRollingRepository(Revision headRevision, int minRetentionCommits,
+                                                   int minRetentionDays) {
+        // TODO(minwoox): provide a way to set different minRetentionCommits and minRetentionDays
+        //                in each repository.
+        if ((minRetentionCommits == 0 && minRetentionDays == 0) ||
+            (minRetentionCommits == Integer.MAX_VALUE && minRetentionDays == Integer.MAX_VALUE)) {
+            // Not enabled.
+            return null;
+        }
+
+        final InternalRepository repo = secondaryRepo != null ? secondaryRepo : primaryRepo;
+        if (exceedsMinRetention(repo, headRevision, minRetentionCommits, minRetentionDays)) {
+            return headRevision;
+        }
+        return null;
+    }
+
+    private static boolean exceedsMinRetention(InternalRepository repo, Revision headRevision,
+                                               int minRetentionCommits, int minRetentionDays) {
+        final Revision firstRevision = repo.firstRevision();
+        if (minRetentionCommits != 0) {
+            if (headRevision.major() - firstRevision.major() <= minRetentionCommits) {
+                return false;
+            }
+        }
+
+        if (minRetentionDays != 0) {
+            final Instant secondCommitCreationTime = repo.secondCommitCreationTimeInstant();
+            return secondCommitCreationTime != null &&
+                   secondCommitCreationTime.isBefore(Instant.now().minus(minRetentionDays, ChronoUnit.DAYS));
+        }
+        return true;
+    }
+
+    @Override
+    public void createRollingRepository(Revision rollingRepositoryInitialRevision,
+                                        int minRetentionCommits, int minRetentionDays) {
+        requireNonNull(rollingRepositoryInitialRevision, "rollingRepositoryInitialRevision");
+        checkState(shouldCreateRollingRepository(rollingRepositoryInitialRevision,
+                                                 minRetentionCommits, minRetentionDays) ==
+                   rollingRepositoryInitialRevision, "aaa");
+
+        final InternalRepository secondaryRepo = this.secondaryRepo;
+        if (secondaryRepo != null) {
+            promoteSecondaryRepo(secondaryRepo, rollingRepositoryInitialRevision);
+        } else {
+            createSecondaryRepo(rollingRepositoryInitialRevision);
+        }
+    }
+
+    private void promoteSecondaryRepo(InternalRepository secondaryRepo,
+                                      Revision rollingRepositoryInitialRevision) {
+        writeLock();
+        try {
+            logger.info("Promoting the secondary repository in {}/{}.", parent.name(), originalRepoName);
```

Comment:
Q) What do you think of validating `secondaryRepo.HEAD == primaryRepo.HEAD` here? (inside the write lock)